### PR TITLE
Use the github api instead of github raw for fetching md

### PR DESCRIPTION
**Pull request by koraa**
Github raw caches the markdown files for a long time; this is
unacceptable when using the helix-bot to clear caches, as it
will result in the stale markdown being transformed again and
being served to the user, instead of proper cache clearing and
serving the *new* markdown version

This also simplifies the markdown fetch code a bit.

### CODE REVIEW

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging f2545f0f5a76ad22a0b701e6a63b2a2f7e6089a7 into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-3066d29968c5415a874d0e6b7824e1d2202bde04)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by tripodsan** [comment]
Why did you remove the timeout checks?

**Comment by koraa** [comment]
I did not; I just set a default timeout. It struct me that there is no real use case to customize those timeouts, though I may be wrong.

I removed the tests because I think we can trust request-promise-native to handle timeouts; no point in duplicating these tests…

File: test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har [hunk 1]
```diff
@@ -8,130 +8,139 @@
...
             {
-              "name": "accept-ranges",
-              "value": "bytes"
+              "name": "x-ratelimit-limit",
+              "value": "60"
             },
             {
-              "name": "date",
-              "value": "Tue, 26 Feb 2019 11:54:16 GMT"
+              "name": "x-ratelimit-remaining",
```
[the remark this answers is not in the document]
Comment (koraa) [reply test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har]: 60 req/minute?
Können wir upgraden? Ansonsten weiß ich nicht wie wir weiter vorgehen sollen. GHs caching auf .raw zerschießt und einmal komplett die cache invalidation und das ist so wie ich das sehe inakzeptabel…

[the remark this answers is not in the document]
Comment (trieloff) [reply test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har]: https://developer.github.com/v3/#rate-limiting

> For unauthenticated requests, the rate limit allows for up to 60 requests per hour. Unauthenticated requests are associated with the originating IP address, and not the user making requests.

[the remark this answers is not in the document]
Comment (trieloff) [reply test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har]: If we manage to pass in a token, we can raise the limit to 5000 requests per hour (per user).

[the remark this answers is not in the document]
Comment (koraa) [reply test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har]: I guess the 5k limit is also per user per gh application? Or per user across al gh applications? I mean the goal is to have it all work from the helix-bot, so we might just generate a token for the user's gh bot installation…
Well, anyways I guess this will linger here for a while until I get the helix bot to a more advanced stage…

File: src/html/fetch-markdown.js [hunk 2]
```diff
@@ -10,102 +10,51 @@
...
 
-  const { REPO_RAW_ROOT: rootPath } = secrets;
-
-  // everything looks good, make the HTTP request
-  const options = {
-    uri: uri(rootPath, owner, repo, ref, path),
-    json: false,
-    timeout,
+  const opt = {
+    url: normalizeURL(`https://api.github.com/repos/${owner}/${repo}/contents/${path}`),
```
Comment (stefan-guggisberg) [inline src/html/fetch-markdown.js]: This breaks local development. GH endpoints should not be hard-coded.  Content might be served from local `git-server`, as e.g. in the following use case:

```bash
hlx demo my-helix-project
cd my-helix-project
hlx up
```

Comment (koraa) [reply src/html/fetch-markdown.js]: Thanks for checking that; I did not test how this behaves with local dev!

This definitely is a problem that needs fixing before this can be merged!
Personally, I don't think just making the domain customizeable  again is a good solution; after all the changes are not just limited to the url; there also are the header changes and there may be more…

So IMO we need a more powerful solution to customize the md fetching step; my first idea would be some `pipe.substitute("fetch", () => custom_impl)` extension point, so we can just provide a custom fetch step in local development…
Together with tooling that makes sure all required repos (as stated in the strain config) are checked out locally, we could get rid of a lot of code and improve the experience for devs working on large projects…

Comment (trieloff) [reply src/html/fetch-markdown.js]: You don't have to substitute, just inject your custom fetch before `fetch`, and then the "real fetch" will be skipped because the content is already there.

File: test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har (hunk 1 above)
[the remark this answers is not in the document]
Comment (trieloff) [reply test/fixtures/Test-non-existing-content_3508203637/Getting-XDM-README-from-wrong-URL-_3748211873/recording.har]: > This means that all OAuth applications authorized by a user share the same quota of 5000 requests per hour when they authenticate with different tokens owned by the same user.

**Comment by kptdobe** [comment]
I would not mix the caching issue which leads to use `api` instead of `raw` and the authorized requests: these are 2 different issues which must be treated separately. Mainly to avoid a big PR with a lot of regressions...

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging 775b08db2076eea36fd54324fbf37c022940e384 into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-d6625ec11681b1810d7847af1b395985b1221ceb)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging b44eb5e090662a9aaaa9a31fab63d56bc6f8f1e4 into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-790f2e7ac2d1d8152b5f01d634d2e1d249244a03)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging 899f498effa13d3aa480a9b83b290a94553263cc into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-2fd54dd7d53d8ef186e6a456d42fa91ef9b9daba)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging fb4c2e6331058b0d2089fa9262e8333c9cb2d5c7 into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-a4ec5471b647cd52db15c8491d03fb793d7a1492)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging 86653f06bbd35782af20669fd57777a2f5dc78e9 into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-3528339a7d43ab5db7d55c4676adc483d0a8925c)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging 539f8c059c310d31e623f915290515937e84f54f into 25af1c340653ececa89c823811de447c2ac501b5 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/helix-pipeline/rev/pr-776e93ab58a65d502850d744b554f9500132759c)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/html/fetch-markdown.js (hunk 2 above)
Comment (tripodsan) [reply src/html/fetch-markdown.js]: but we pass the `REPO_API_ROOT` which is adjusted to the correct github server during local development. for example:

```
"action": {
  "secrets": {
    "REPO_RAW_ROOT": "http://127.0.0.1:53571/raw/",
    "REPO_API_ROOT": "http://127.0.0.1:53571/api/",
    ...
  },
  "request": {
    "params": {
      "owner": "helix",
      "repo": "github.com--tripodsan--hlxtest",
      "ref": "master",
      "path": "/index.md",
      ...
    },
}
```
So I think if you just use `action.secrets.REPO_API_ROOT`, you're good.

**Comment by tripodsan** [comment]
> I would not mix the caching issue which leads to use `api` instead of `raw` and the authorized requests: these are 2 different issues which must be treated separately. Mainly to avoid a big PR with a lot of regressions...

right: there are 2 points here:

1. use API or RAW to fetch the markdown (and statics, btw)
2. authenticate the request or not.

(2) is required for private repository access. this is irrespective which endpoints to use.
(1) leads to better cache freshness on github's side. but comes with the drawback, that unauthenticated requests have a limit of 60/s. 

What makes @koraa's problem worse is, that the simple-setup renders everyones content with the same user, ever if authenticated. 

What about having a hybrid solution, with fastly acting as "content-cache"? :
- request content through fastly: `https://content.project-helix.io/tripodsan/hlxtest/index.md`
- use github API as backend, authenticated if possible. calculate a surrogate key based `owner/repo`.
- when invalidate the cache, use the `owner/repo` key

this give the advantage:
- that we control the content cache completely
- we can set max-age to infinity, since we always know what to purge
- very few requests to API

---

**BTW: we should move this discussion to an issue. IMO, new features/fixes should always start as issues. **

File: src/html/fetch-markdown.js [hunk 3]
```diff
@@ -10,102 +10,57 @@
...
-    const response = await client(options);
-    return {
-      content: {
-        body: response,
-        sources: [...sources, options.uri],
-      },
-    };
-  } catch (e) {
-    if (e && e.statusCode && e.statusCode === 404) {
-      return bail(logger, `Could not find Markdown at ${options.uri}`, e, 404);
```
Comment (tripodsan) [inline src/html/fetch-markdown.js]: why did you remove all this?

File: test/testFetchMarkdown.js [hunk 4]
```diff
@@ -23,103 +23,6 @@ const logger = Logger.getTestLogger({
   level: 'info',
 });
 
-describe('Test URI parsing and construction', () => {
```
Comment (tripodsan) [inline test/testFetchMarkdown.js]: why did you remove all tests?

File: src/html/fetch-markdown.js [hunk 5]
```diff
@@ -10,102 +10,57 @@
...
-  if (!ref) {
-    logger.warn(`Recoverable error: no ref given for ${repo}/${owner}.git${path}, falling back to master`);
-    ref = 'master';
-  }
+  logger.info(`C`);
+  const opt = {
+    url: normalizeURL(`https://api.github.com/repos/${owner}/${repo}/contents/${path}`),
+    qs: { ref },
+    headers: {
+      Accept: 'application/vnd.github.VERSION.raw"',
```
Comment (trieloff) [inline src/html/fetch-markdown.js]:
```suggestion
      Accept: 'application/vnd.github.VERSION.raw',
```

I think you have a superfluous `"` in there.

**Comment by trieloff** [comment]
Closing stale PR.
